### ISSUE TITLE
storage: command queue should support "deferred cancellation"

### ISSUE THREAD

**Issue by tamird** [cockroachdb/cockroach#9237]
From discussion in #9177:

@petermattis:

> Doing this cleanup asynchronously like this feels like a failure of the CommandQueue API. I think it would be cleaner/better if the added command could be marked for removal by the CommandQueue when its dependencies finish. I don't see a clear way to do that, but this is only a drive-by comment.

The command queue assumes that commands don't drop out before their prerequisites. Consider the following scenario:
- C1 enters the command queue
- C2 (depends on C1, covers it entirely) enters the command queue
- C3 enters the command queue, depends on C2 (but not C1, because C1 is covered by C2)
- C2's context is cancelled

What happens next changed in #9177:

Prior to that PR, C2's caller would block until C1 completed, and then would remove C2 from the command queue (without executing it).
After that PR, C2's caller spawns a goroutine which blocks until C1 completes, and then removes C2 from the command queue. C2's caller returns immediately.

The discussion in #9177 suggests that it is possible to avoid the extra goroutine by offloading the cleanup work to the abandoned commands' dependents.
